packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,263 @@
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type Subscribers = Set<Subscriber>;
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscribers: Subscribers | null = null;
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+  if (enableInteractionTrackingObserver) {
+    subscribers = new Set();
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {interactionsRef as __interactionsRef, subscribers as __subscribers};
+
+export function clear(callback: Function): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+  ((interactionsRef: any): InteractionsRef).current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    ((subscribers: any): Subscribers).add(subscriber);
+  }
+}
+
+export function track(
+  name: string,
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const interaction: Interaction = {
+    id: interactionIDCounter++,
+    name,
+    timestamp: now(),
+  };
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+
+  // Tracked interactions should stack/accumulate.
+  // To do that, clone the current interactions array.
+  // The previous interactions array will be restored upon completion.
+  const interactions = new Set(prevInteractions);
+  interactions.add(interaction);
+  ((interactionsRef: any): InteractionsRef).current = interactions;
+
+  try {
+    if (enableInteractionTrackingObserver) {
+      interaction.__count = 1;
+
+      ((subscribers: any): Subscribers).forEach(subscriber => {
+        subscriber.onInteractionTracked(interaction);
```

Comment:
Do we have any guarantees around what happens if one of these methods throws? Do we assume subscribers themselves are crash-free? (Which we could assume — but I think it's important that we communicate this explicitly, especially if a broken subscriber would bring down the product code too.)

If we assume subscribers are crash-free we could move `try` closer to `callback`. If we don't assume that, we should probably isolate failure in subscribers from calling the `callback`?

Reply:
I tried to write the code in a way that guarded against this– e.g. by restoring the interactions set before calling handlers.

I will add some error handling tests for the observers though. I should have done that before.

Reply:
To be clear, what I'm concerned about is that if subscribers fail, we don't call the actual `callback` function, which from my understanding, would break product code (e.g. we wouldn't call the event handler implementation).

Reply:
Right. I am leaning towards being strict with subscribers though. If they error, it's like an error in the callback. We won't leave the module in a corrupted state, but we won't try to run code like nothing happened either.

Reply:
I've added tests and tightened up the error handling in b8287cf.

I'm still on the fence about whether it's worth adding additional complexity to e.g. catch an error in on subscriber, but still call other subscribers.